### ISSUE TITLE
Different Sample Weights for Different Output Nodes

### ISSUE THREAD

**Issue by skarakulak** [fchollet/keras#549]
I'm trying to get acquainted with keras, and if I understood it correctly, sample_weight argument in the fit() takes arrays, and each element of this array is multiplied with the loss value of the corresponding sample. What I am interested in is differentiating these weights for each output node. Can this be done in some way?

To give a background to this question, I am working on a classification problem, where there are 20 output nodes. And for each output node ```y_m``` there is a input node ```x_k```, such that if ```x_k``` is equal to 0, then ```y_m``` is 0. ```y_m``` can only be 1 if ```x_k``` is 1 (those two variables are binary by the way). So it would be fitting to use something like ```sample_weight = x[,1:20]``` as an argument.

**Comment by fchollet** [comment]
I think what you want here is a custom objective function that would implement this. Check out `keras/objectives.py` for some examples.

**Comment by skarakulak** [comment]
The functions in the ```objectives.py``` has only ```(y_true, y_pred)``` as arguments. Can I use ```x``` values here?

**Comment by fchollet** [comment]
No, though you could include this information as part of the target `y_true`. Targets and predictions may have different dimensionality if the objective function supports it.

**Comment by skarakulak** [comment]
Hmm that makes sense. So I would have 40 output nodes, 20 of those would be the predictions, and the other 20 are the input values which affect predictions in question. Loss would be calculated on ```y_true_predictions * y_true_inputs```, where ```y_true_inputs``` values are one if the loss would be calculated, and zero otherwise.

Thanks for the insight:) Cheers

**Comment by giamba92** [comment]
I need the same feature and I tried to implement what @fchollet and @skarakulak recommended, could you guys please provide feedback about the implementation? It seems to be working fine, I use y_true as a horiz concatenation of the targets and the mask

`
 # the mask, in this case with all ones
sample_weights_train = np.ones((Y_train.shape[0],Y_train.shape[1]))

# custom loss 
from keras import backend as K
def node_weighted_binary_crossentropy(y_true, y_pred):
    y_t = y_true[:,:y_true.shape[1]//2]*y_true[:,y_true.shape[1]//2:]
    y_p = y_pred*y_true[:,y_true.shape[1]//2:]
    return K.mean(K.binary_crossentropy(y_t, y_p), axis=-1)

model.fit(X_train, np.hstack((Y_train, sample_weights_train)), batch_size=batch_size, nb_epoch=nb_epoch, verbose=2)

`
Would there be any less hacky way of doing the same in Keras 1.0?
When I set all the mask values to 1 the training curves look a little different, so there might be something wrong

**Comment by curiale** [comment]
Hi, I need to do something similar, but, I'm using real data augmentation. So, It is not possible to include the weight information (x value according to @skarakulak) into the Y_true value because the data augmentation procedure is modifying the true value according to the deformation used for extending the data set. Any idea about how I could introduce weights to the objective function?
